doc/remailers.rst

```diff
@@ -0,0 +1,125 @@
+
+Autocrypt-enabled Remailer
+----
+
+An Autocrypt-enabled remailer maintains an Autocrypt key of its own. During
+regular operation, it collects Autocrypt-headers of all list participants via
+incoming mail, to keep track of their keys. Then, depending on the availability
+of encryption keys for participants, the remailer injects its own
+Autocrypt-headers into outgoing mail to signal availability of encryption.
+
+More specifically, an Autocrypt-enabled mailing list adds three behaviors to its
+remailing logic:
+
+1) For all incoming messages, keep track of Autocrypt-headers. This mechanism is
+   the same that is used by clients, and requires no modifications.
+2) Inject an Autocrypt-header into outgoing messages. This header serves the
+   same purpose as usual, but adds some details for the mailing list use case.
+   The header might only be injected under certain conditions, e.g. if
+   encryption keys are available for a certain number of subscribers.
+3) When receiving an encrypted message, decrypt it and re-encrypt for individual
+   delivery to each Autocrypt-enabled subscriber. Subscribers for whom no keys
+   are available do not receive the message, but optionally a notification of
+   this circumstance (see below).
+
+The presented mechanism works well for relatively small groups of mostly active
+and Autocrypt-enabled subscribers, but becomes increasingly ineffective with
+larger numbers of passive or non Autocrypt-enabled subscribers. As with any
+encrypted group communication, the benefits of end-to-end encryption also dilute
+with larger numbers of recipients.
+
+For compatibility with clients that are not Autocrypt-capable, the remailer
+software might want to offer manual configuration of keys, e.g. via a web
+interface.
+
+Injecting Reamiler Autocrypt-Headers
+~~~~~
+
+To inform subscribers of its key, the remailer injects an Autocrypt-header into
+outgoing messages. The header format is the same as usual, where the ``addr``
+attribute is the address of the list, and ``keydata`` is their Autocrypt key.
+
+In addition to those two attributes, the remailer SHOULD add the
+``_list-recommendation`` attribute, and MAY add ``_list-recipients`` attribute.
+Both of these attributes are exclusive for use by remailers, and MUST NOT be
+used by regular mailbox addresses. Both of these attributes are non-critical
+(starting with an underscore), which will cause an Autocrypt-capable client that
+does not handle these attributes to skip over them, falling back to the usual
+Autocrypt behavior.
+
+For example::
+
+    Autocrypt: addr=post@list.example; _list-recommendation=mandatory; keydata=XXX
+    Autocrypt: addr=post@list.example; _list-recommendation=available; _list-recipients=7/15; keydata=XXX
+
+The ``_list-recommendation`` attribute SHOULD be included in an Autocrypt header
+sent by a remailer. Its value may be either ``available`` or ``mandatory``. If
+the value is set to ``mandatory``, the remailer requires encryption of every
```

Comment:
Note that this introduces a `mandatory` mode to Autocrypt. I consider this a notably different use case than a unilateral preference for encryption by default, because it implies that messages that aren't encrypted really strictly are not going to be delivered. There should probably be an override, but it's a nice usability feature to have an immediate warning for fully encrypted lists.

Reply:
do we want to include this here?  It's conceivable that this "mandatory" mode could be relevant for non-list mailboxes as well, where cleartext messages are refused delivery.  if so, it'd be strange to have two ways to express it.

I recommend splitting the idea of ``mandatory`` out into a separate/orthogonal discussion, and removing `_list-recommndation`` from this draft.

Reply:
I don't think it would be strange at all, because there is value in displaying a specific message to the user about the actual case. Even if the technical implication is the same or similar, it feels like a good idea to me to keep the use cases apart on the protocol layer, otherwise implementations *will* conflate them. "This recipient is a fully encrypted mailing list, unencrypted messages may not be delivered."

This is also the reason for the `MUST NOT` wrt using those attributes for non-mailinglist addresses.

Reply:
I'm still not convinced, sorry.  two functionally equivalent specs *will* be conflated by implementations even with a `MUST NOT` that tries to scare the implementers off.

and `_list-recommendation` isn't the only way to flag the list as a mailing list -- we can use `_list-recipients` (or a critical `list-recipients`, if we want to require the MUA to be able to know about this breach of e2e to present it as encryption-capable to the user).

Bundling the concept of "cleartext mail will not be delivered" with "you are sending to a mailing list" really doesn't seem like the right way to go.

Reply:
The `mandatory` bit was not included in my first draft, I added it based on feedback by @ge-fa for the specific use case of Schleuder. Perhaps he can also weigh in on this.

> Bundling the concept of "cleartext mail will not be delivered" with "you are sending to a mailing list" really doesn't seem like the right way to go.

There is an important difference here: For mailing lists, messages *actually won't be delivered*. This is not the case for regular mailboxes, "I will disregard plaintext messages you send to me" is not something that actually works with a human backend.

I do think the `mandatory` mode is useful, but I'm still sceptical if we should design this for a more general use case that we don't have a clear picture of and can't really reason about. Do you have another use case in mind?

A more general take leads to `prefer-encrypt=mandatory`, but honestly the abuse potential of such a setting is so high that I'd rather not have it in Autocrypt.